### DISCUSSION TITLE
CLI changes to follow best SW engineering practices and be backward compatible

### DISCUSSION THREAD

**Discussion by dmfiore** [openthread/openthread#7659] [Q&A] [answered]
Hi,

I don't know if this topic has been raised.  I was wondering what the current OT policy is on backward compatibility.
I did a couple of searches on 'backward compatibility' and 'CLI changes' but didn't find anything.

As a SQA engineer, I think it is important that changes to OT do not break what exists. 
For CLI syntax changes, this can cause a lot of disruption for many engineers developing test automation over the evolution of this protocol.

I am new to this forum, but not the protocol.  @jwhui is there a best practices policy in this respect for OT and backward compatibility?
Thank you,
-D

**Comment by jwhui** [comment, marked as answer]
@dmfiore , thanks for raising this question.

OpenThread does not currently have any policy for the CLI. Historically, the CLI was primarily used as a simple mechanism for quick testing during software and development. Given that the CLI operates on resource-constrained devices, we wanted to minimize resource requirements for the CLI. At the same time, given that Thread is a relatively new technology, we did not want to constrain ourselves to design decisions that were made in the past as we obtain new learnings and insights moving forward.

That said, we could be more diligent in maintaining a CHANGES list for the CLI. Similar to what we do with the [OpenThread API](https://openthread.io/reference/api-updates). Would that help?

**Reply by dmfiore** [reply]
Thanks for the quick reply.  I understand, CLI is just a testing layer not used in production.  I would argue it is more than that and the first mechanism test engineers usually go to for proof of concept and then build upon that.
Yes, I definitely think more detailed changes to doc always help, but ideally there would be a deprecation process. 

I also think we could do a few things even better with some more guidance and validation.

For changes that break existing, couldn't we just add an alias, a stub, something that calls the same function just without the invalid arg? Maybe just warn and report deprecation, do not invalidate the syntax that was previous valid?  I think the value in this managed at the source might worth it for the growing number of engineers testing and setting up validation automation.

Are there any CLI syntax validation checks in the current PR validation?
Thanks again,
-D

PS To be clear, I certainly don't want to constrain ourselves to design decisions in the past, but we should be accountable for it, and maybe grab some low hanging fruit to do both ?

**Reply by jwhui** [reply]
My primary concern is that OpenThread needs to support devices that only have resources to support several hundred kilobytes of compiled code. The CLI component is already a relatively large component in OpenThread. If we also had to support all previous syntax that was ever implemented, I'm not sure that would be supportable on the class of devices we are targeting. To help guide the discussion, can you help provide some examples of CLI changes that were problematic?

FWIW, we do not have a backward-compatibility policy on the OpenThread API. What we found is that OpenThread is typically an ingredient component in larger systems and SDKs. As a result, the vast majority of developers are developing to APIs of those larger systems and SDKs, rather than OpenThread itself. For this reason, having a stable API for OpenThread has not been as much of a concern as it has for other software components.

**Reply by dmfiore** [reply]
I understand the constraints, I'm curious if it would be possible to have a 'minimal CLI' vs 'full CLI' build ?  Where 'full' is supporting the all previous syntax ?  And 'min' can even not include that, but also trim even more from full set that we as a group don't seem to need?

An example recent change, is the removal of a datasetsecuritypolicy flag, I think it was -b.  Our SQA relies upon CLI expect based framework.  A change in CLI is a break that the abstracted framework needs to address over different internal branches as the github change is populated through various branches over time.  
There's been a couple example like this over the last year or so; so not frequent, just a pain (and cycles).

I am surprised there is no backward-compatibility policy in place.  This is something that would have never passed within our teams, and does not.  I am surprised that other companies are okay with no backward compatibility support or validation.

Are there any plans or discussions promoting these above discussions in any way?
Thanks,
-D

**Reply by dmfiore** [reply]
To ask further, beyond CLI, is there ANY backward compatibility policy for APIs ?  ( I thought I heard no here and just wanted to clarify).

**Reply by jwhui** [reply]
As previously mentioned, we do not have a formal backward compatibility policy for both the CLI and API. For the API, we do maintain a [OPENTHREAD_API_VERSION](https://github.com/openthread/openthread/blob/f3774e3368a5656ba2aa994276a1f1dca6ed734c/include/openthread/instance.h#L56), which must increment any time the API changes in any way. This allows integrators to develop code that can build against various OpenThread commits. That said, we do review changes to the interfaces and try not to make gratuitous changes.

For the specific example you mentioned, the `b` flag was actually removed from the Thread Specification. In this particular case, I think it is better for the developer to understand up front that the `b` flag is no longer a valid option from a standards perspective.

So far, many developers leveraging Thread do not actually develop directly against the OpenThread API or CLI. They typically use higher-layer abstractions provided by other SDKs (typically from silicon vendors). As a result, the various silicon vendors have taken on the responsibility to ensure their customers have a good experience.

That said, this is an open discussion forum. If others have opinions on this topic, please comment here.

**Reply by dmfiore** [reply]
Thank you for the quick response and dialogue. 
A point I was trying to push is yes, a flag became invalid.
Can that be handled gracefully vs hard fault ?  For something that once was valid.
While it is CLI, it is also attached to the API.  I do not know if that translates to an invalid value over the air.  But it could possibly happen under these conditions of validation.
-D

**Reply by abtink** [reply]
@dmfiore, let me share some of the policies and guidelines we have in OT (related to compatibility of different components):

- **NCP spinel definition**: We require spinel definitions to be backward compatible (in the sense that older host side driver should be able to work with new NCP build). This is enforced by ensuring definitions are always additive (new fields can be added to existing spinel properties, but previously defined fields and their meaning cannot change. 
- **RCP spinel definition**: This is more complex. We have two version numbers that are tracked, you can see more details here: [spinel.h](https://github.com/openthread/openthread/blob/f3774e3368a5656ba2aa994276a1f1dca6ed734c/src/lib/spinel/spinel.h#L293-L330).
- **OT_CONFIG**: Existing `OT_CONFIG` config definitions are not changed often, but if they are, any removed or renamed config is tracked in [`openthread-core-config-check.h`](https://github.com/openthread/openthread/blob/f3774e3368a5656ba2aa994276a1f1dca6ed734c/src/core/config/openthread-core-config-check.h) (which causes build failure in case any are used by a project/platform).
- **OT APIs** :As Jonathan mentioned, the OT API changes are tracked by `OPENTHREAD_API_VERSION` (any change in any of public OT header files requires a `OT_API_VERSION` increase even an update to comments). This allow users of OT to track and know which API version they have tested and used. The update of `OPENTHREAD_API_VERSION` is checked and enforced as part of the GitHub actions.

Related to OT APIs:
- The general guideline is to avoid removing/changing existing APIs (in particular well-established APIs). 
- New features and behaviors are often added as new APIs and new definitions. 
- CLI is to a great extent mirror the OT API and therefore follows the same guideline.
- However, as you know, this is not always possible and there are situations where we would have changes in existing APIs (and/or related CLI) that are more fundamental. The general goal is however to minimize such changes.
- I understand the desire for backward compatibility (and the pain of adopting incompatible changes)
- But I strongly believe **in the cases where we need/have such changes (i.e., when a behavior is fundamentally changed), it is much better to break the compatibility fully (i.e., the API/definitions and related CLI should differ)** 
- **In my view, in such cases, trying to keep API/CLI backward compatible actually can do more harm since any code that relies on the changed behavior may not notice the change and assume it is still provided.** 
- I think what you mention is a good example of this case (which removed the `mBeaconsEnabled` in security policy and all its related behavior including CLI changes).

**Reply by dmfiore** [reply]
Perfect.  Thank you.
I don't outright disagree, I appreciate the dialogue.   
That said, let me digest a bit on these responses,
=D

**Reply by jwhui** [reply]
> A point I was trying to push is yes, a flag became invalid. Can that be handled gracefully vs hard fault ? For something that once was valid. While it is CLI, it is also attached to the API. I do not know if that translates to an invalid value over the air. But it could possibly happen under these conditions of validation.

I believe the CLI returns an error to indicate that the `b` flag is no longer valid and does not generate a hard fault. In this particular case, I believe returning an error is the proper behavior since users of the `b` flag should be aware that this functionality no longer exists in the Thread standard.

Ideally, any test infrastructure leveraging the CLI should check for and report CLI errors. If that is the case, I believe this issue could have been addressed with relatively little effort. This is how we implement all of the OpenThread continuous integration test scripts.

**Reply by dmfiore** [reply]
Agreed on the test infrastructure, our teams continually move in that direction upon changes of either env, cli, or other, yes.

There's one thing here running around my head...  
What happens when a version 1.1 in the field talks to a version 3.1 in the field for which multiple API changes have been made over time?
How is interoperability discussed and handled for these cases ?  Do we test any version interoperability with OT?
I appreciate the discussions, and we can take offline, though I think this is a good thread to discuss intent and some philosophy?
-D

**Reply by jwhui** [reply]
Thread Protocol interoperability is very different from API and CLI backward compatibility. With the Thread Protocol, we need to make sure that devices can continue to communicate over the air. The Thread Group's Certification program tests that a given device is backward compatible with all prior Thread protocol versions. The OpenThread project also includes [continuous integration tests for backwards compatibility](https://github.com/openthread/openthread/blob/f3774e3368a5656ba2aa994276a1f1dca6ed734c/.github/workflows/simulation-1.2.yml#L53).

**Reply by simonlingoogle** [reply]
Agree with @jwhui and @abtink on the general principles for dealing with CLI compatibilities. 

I had several issues regarding CLI compatibility, too. 
These are the general strategies I have been using to address CLI compatibility issues:

- Check the [API version](https://github.com/openthread/openthread/blob/main/src/cli/README.md#version-api) and choose different commands accordingly. 
- Check if a command exists before running it. (e.g. `masterkey`)

Overall, I think we are already trying to make the CLI commands and output stable. And it's not difficult to handle when it's not.

**Reply by dmfiore** [reply]
Thanks for all the feedback!   Good info to absorb.
-D
